crates/telio-traversal/src/endpoint_providers/upnp/mod.rs

```diff
@@ -35,35 +41,71 @@ const GET_INTERFACE_TIMEOUT: u64 = 2;
 
 type Result<T> = std::result::Result<T, Error>;
 
-#[async_trait]
 #[cfg_attr(test, automock)]
+#[async_trait]
 pub trait UpnpEpCommands: Send + Sync + Default + 'static {
-    fn check_endpoint_routes(&self, proxy_port: u16, wg_port: u16) -> Result<bool>;
-    fn add_any_endpoint_routes(
+    async fn check_endpoint_routes(&self, proxy_port: u16, wg_port: u16) -> Result<bool>;
+    async fn add_any_endpoint_routes(
         &self,
         ip_addr: Ipv4Addr,
         proxy_port: u16,
         wg_port: u16,
     ) -> Result<(u16, u16)>;
-    fn extend_endpoint_duration(
+    async fn extend_endpoint_duration(
         &self,
         ip_addr: Ipv4Addr,
         proxy_port: PortMapping,
         wg_port: PortMapping,
     ) -> Result<()>;
-    fn delete_endpoint_routes(&self, proxy_port: u16, wg_port: u16) -> Result<()>;
-    fn get_igd_gateway(&mut self) -> Result<()>;
-    fn get_external_ip(&self) -> Result<Ipv4Addr>;
+    async fn delete_endpoint_routes(&self, proxy_port: u16, wg_port: u16) -> Result<()>;
+    async fn get_igd_gateway(&mut self) -> Result<()>;
+    async fn get_external_ip(&self) -> Result<Ipv4Addr>;
 }
 
-#[derive(Default, Debug, Clone)]
+#[derive(Debug)]
 pub struct IgdGateway {
     gw: Option<Gateway>,
+    rx: mpsc::Receiver<std::result::Result<Gateway, SearchError>>,
+    tx: mpsc::Sender<std::result::Result<Gateway, SearchError>>,
+    search_handle: Option<tokio::task::JoinHandle<()>>,
+}
+
+impl Default for IgdGateway {
+    fn default() -> Self {
+        Self::new(true)
+    }
+}
+
+impl IgdGateway {
+    fn new(auto_search: bool) -> Self {
+        let (tx, rx) = mpsc::channel::<std::result::Result<Gateway, SearchError>>(1);
+
+        let mut igd_gateway = Self {
+            gw: None,
+            rx,
+            tx,
+            search_handle: None,
+        };
+
+        if auto_search {
+            igd_gateway.start_gw_search();
+        }
+
+        igd_gateway
+    }
+
+    fn start_gw_search(&mut self) {
+        let tx = self.tx.clone();
+        self.search_handle = Some(tokio::task::spawn(async move {
+            let result = Box::pin(search_gateway(Default::default())).await;
+            let _ = tx.send(result).await;
```

Comment:
Look to me like you can simply use task return instead of channel.
Then rewriting to something like this:
```rust
    async fn get_igd_gateway(&mut self) -> Result<()> {
        match &mut self.search_handle {
            Some(handle) if handle.is_finished() => {
                self.gw = handle.await.ok().and_then(std::result::Result::ok);
            }
            None => self.start_gw_search(),
            _ => (),
        };

        // Gateway search in progress, check again on the following iteration
        Err(Error::NoIGDGateway)
    }
```
Maybe adding warn log's for failure case